File: agent/collectors/software.py
import platform
import subprocess
import json

def get_software_list():
    """Fast Registry-based Software Scan + Custom Folder Detections"""
    software = []
    if platform.system() == "Windows":
        # 1. Custom Detections (e.g., BMS HOSxP XE4)
        try:
            import glob
            import os
            users_dir = os.environ.get("SystemDrive", "C:") + "\\Users"
            search_pattern = os.path.join(users_dir, "*", "AppData", "Roaming", "BMS", "HOSxPXE4")
            found_paths = glob.glob(search_pattern)
            for path in found_paths:
                if os.path.isdir(path):
                    software.append({
                        "name": "HOSxP XE4",
                        "version": "N/A",
                        "publisher": "BMS (Bangkok Medical Software)",
                        "install_path": path,
                        "install_date": ""
                    })
        except Exception:
            pass

        # 2. Registry-based Software Scan
        try:
            cmd = [
                "powershell",
                "-NoProfile",
                "-ExecutionPolicy",
                "Bypass",
                "-Command",
                "Get-ItemProperty "
                "HKLM:\\Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\*, "
                "HKLM:\\Software\\WOW6432Node\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\*, "
                "HKCU:\\Software\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\* "
                "| Select-Object DisplayName,DisplayVersion,Publisher,InstallLocation,InstallDate "
                "| ConvertTo-Json -Depth 3",
            ]
            result = subprocess.run(cmd, capture_output=True, text=True, shell=False, timeout=60)
            if result.stdout:
                data = json.loads(result.stdout)
                if isinstance(data, dict): data = [data]
                seen = set()
                # Initialize seen with already detected software
                for sw in software:
                    seen.add((sw["name"].lower(), sw["version"].lower(), sw["publisher"].lower()))
                for item in data:
                    name = (item.get("DisplayName") or "").strip()
                    if not name:
                        continue
                    version = (item.get("DisplayVersion") or "").strip() or "N/A"
                    publisher = (item.get("Publisher") or "").strip()
                    install_path = (item.get("InstallLocation") or "").strip() or "N/A"
                    install_date = (item.get("InstallDate") or "").strip()

                    key = (name.lower(), version.lower(), publisher.lower())
                    if key in seen:
                        continue
                    seen.add(key)

                    software.append({
                        "name": name,
                        "version": version,
                        "publisher": publisher,
                        "install_path": install_path,
                        "install_date": install_date,
                    })
        except:
            pass
    return software

File: go_agent/collectors/software.go
package collectors

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"golang.org/x/sys/windows/registry"
)

type SoftwareEntry struct {
	Name        string `json:"name"`
	Version     string `json:"version"`
	Publisher   string `json:"publisher"`
	InstallPath string `json:"install_path"`
}

// GetSoftwareList scans installed programs from registry + custom HOSxP detection
func GetSoftwareList() []SoftwareEntry {
	seen := map[string]bool{}
	var list []SoftwareEntry

	// 1. Custom detection: HOSxP XE4 via AppData scan
	if usersDir := os.Getenv("SystemDrive"); usersDir != "" {
		pattern := filepath.Join(usersDir+"\\Users", "*", "AppData", "Roaming", "BMS", "HOSxPXE4")
		matches, _ := filepath.Glob(pattern)
		for _, p := range matches {
			info, err := os.Stat(p)
			if err == nil && info.IsDir() {
				key := "hosxp xe4"
				if !seen[key] {
					seen[key] = true
					list = append(list, SoftwareEntry{
						Name:        "HOSxP XE4",
						Version:     "N/A",
						Publisher:   "BMS (Bangkok Medical Software)",
						InstallPath: p,
					})
				}
			}
		}
	}

	// 2. Registry scan from all standard Uninstall paths
	paths := []struct {
		root registry.Key
		path string
	}{
		{registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall`},
		{registry.LOCAL_MACHINE, `SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall`},
		{registry.CURRENT_USER, `SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall`},
	}

	for _, p := range paths {
		k, err := registry.OpenKey(p.root, p.path, registry.READ)
		if err != nil {
			continue
		}
		subkeys, err := k.ReadSubKeyNames(-1)
		k.Close()
		if err != nil {
			continue
		}

		for _, sub := range subkeys {
			subPath := p.path + `\` + sub
			sk, err := registry.OpenKey(p.root, subPath, registry.READ)
			if err != nil {
				continue
			}

			name, _, _ := sk.GetStringValue("DisplayName")
			version, _, _ := sk.GetStringValue("DisplayVersion")
			publisher, _, _ := sk.GetStringValue("Publisher")
			installPath, _, _ := sk.GetStringValue("InstallLocation")
			sk.Close()

			name = strings.TrimSpace(name)
			if name == "" {
				continue
			}
			version = strings.TrimSpace(version)
			publisher = strings.TrimSpace(publisher)
			installPath = strings.TrimSpace(installPath)
			if version == "" {
				version = "N/A"
			}
			if installPath == "" {
				installPath = "N/A"
			}

			dupKey := strings.ToLower(fmt.Sprintf("%s|%s|%s", name, version, publisher))
			if seen[dupKey] {
				continue
			}
			seen[dupKey] = true
			list = append(list, SoftwareEntry{
				Name:        name,
				Version:     version,
				Publisher:   publisher,
				InstallPath: installPath,
			})
		}
	}

	return list
}
